### PR TITLE
Support for Code Signing For AWS Lambda

### PR DESCRIPTION
**Pull request by santiagocardenas**
Add support in Lambda Layers for Code Signing for AWS Lambda
Add new Code Signing Config data source/resource and support in Functions for Code Signing for AWS Lambda.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
FEATURES

    New Data Source: aws_lambda_code_signing_config
    New Resource: aws_lambda_code_signing_config

ENHANCEMENTS

    data-source/aws_lambda_function: Add `signing_profile_version_arn` and `signing_job_arn` attributes
    data-source/aws_lambda_layer_version: Add `signing_profile_version_arn` and `signing_job_arn` attributes
    resource/aws_lambda_function: Add `signing_profile_version_arn` and `signing_job_arn` attributes
    resource/aws_lambda_function: Add `code_signing_config_arn` argument
    resource/aws_lambda_function: Add error checking / handling on the Set() calls
    resource/aws_lambda_layer_version: Add `signing_profile_version_arn` and `signing_job_arn` attributes

```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccDataSourceAWSLambdaCodeSigningConfig*'

--- PASS: TestAccDataSourceAWSLambdaCodeSigningConfig_basic (14.49s)
--- PASS: TestAccDataSourceAWSLambdaCodeSigningConfig_Description (14.51s)
--- PASS: TestAccDataSourceAWSLambdaCodeSigningConfig_PolicyConfigId (14.53s)


$ make testacc TESTARGS='-run=TestAccAWSLambdaCodeSigningConfig*'

--- PASS: TestAccAWSLambdaCodeSigningConfig_basic (14.13s)
--- PASS: TestAccAWSLambdaCodeSigningConfig_UpdatePublishers (23.27s)
--- PASS: TestAccAWSLambdaCodeSigningConfig_UpdatePolicy (24.93s)


$ make testacc TESTARGS='-run=TestAccAWSLambdaFunction_codeSigningConfig'

--- PASS: TestAccAWSLambdaFunction_codeSigningConfig (62.95s)

$ make testacc TESTARGS='-run=TestAccDataSourceAWSLambdaFunction*'

...

$ make testacc TESTARGS='-run=TestAccAWSLambdaFunction*'

...

$ make testacc TESTARGS='-run=TestAccDataSourceAWSLambdaLayerVersion*'

...

$ make testacc TESTARGS='-run=TestAccAWSLambdaLayerVersion*'

...
```

### CODE REVIEW

File: aws/data_source_aws_lambda_code_signing_config_test.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+  policies {
+    untrusted_artifact_on_deployment = "Warn"
+  }
+}
+
+data "aws_lambda_code_signing_config" "test" {
+  arn = aws_lambda_code_signing_config.test.arn
+}
+`
```
Comment (anGie44) [inline aws/data_source_aws_lambda_code_signing_config_test.go:98]: To remove hardcoded ARN e.g. with 1 signing profile resource
```suggestion
resource "aws_signer_signing_profile" "test" { 
    platform_id = "AWSLambda-SHA384-ECDSA"
}
resource "aws_lambda_code_signing_config" "test" {
  allowed_publishers {
    signing_profile_version_arns = ["${aws_signer_signing_profile.test.version_arn}"]
  }

  policies {
    untrusted_artifact_on_deployment = "Warn"
  }
}

data "aws_lambda_code_signing_config" "test" {
  arn = aws_lambda_code_signing_config.test.arn
}
`
```

File: aws/data_source_aws_lambda_code_signing_config_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      "arn:aws:signer:us-east-1:123456789012:signing-profiles/my_profile1/abcde12345",
+      "arn:aws:signer:us-east-1:123456789012:signing-profiles/my_profile2/abcde12345"
+    ]
+  }
+}
+
+data "aws_lambda_code_signing_config" "test" {
+  arn = aws_lambda_code_signing_config.test.arn
+}
+`
```
Comment (anGie44) [inline aws/data_source_aws_lambda_code_signing_config_test.go:79]:
```suggestion
resource "aws_signer_signing_profile" "test" {
  platform_id = "AWSLambda-SHA384-ECDSA"
}

resource "aws_lambda_code_signing_config" "test" {
  allowed_publishers {
    signing_profile_version_arns = ["${aws_signer_signing_profile.test.version_arn}"]
  }
}

data "aws_lambda_code_signing_config" "test" {
  arn = aws_lambda_code_signing_config.test.arn
}
`
```

Comment (santiagocardenas) [reply aws/data_source_aws_lambda_code_signing_config_test.go]: Updated.

File: aws/data_source_aws_lambda_function.go [hunk 3]
```diff
@@ -281,6 +303,23 @@ func dataSourceAwsLambdaFunctionRead(d *schema.ResourceData, meta interface{}) e
...
 
+	// Get Code Signing Config Output
+	// If code signing config output exists, set it to that value, otherwise set it empty.
+	codeSigningConfigInput := &lambda.GetFunctionCodeSigningConfigInput{
+		FunctionName: aws.String(d.Get("function_name").(string)),
+	}
+
+	getCodeSigningConfigOutput, err := conn.GetFunctionCodeSigningConfig(codeSigningConfigInput)
+	if err != nil {
+		return err
```
Comment (bflad) [inline aws/data_source_aws_lambda_function.go:314]: Error messaging context for operators and code maintainers:

```suggestion
		return fmt.Errorf("error getting Lambda Function (%s) Code Signing Config: %w", aws.StringValue(function.FunctionName), err)
```

File: website/docs/d/lambda_code_signing_config.html.markdown [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `allowed_publishers` - List of allowed publishers as signing profiles for this code signing configuration.
+* `config_id` - Unique identifier for the code signing configuration.
+* `description` - Code signing configuration description.
+* `last_modified` - The date and time that the code signing configuration was last modified.
+* `policies` - The code signing policy controls the validation failure action for signature mismatch or expiry.
```
Comment (anGie44) [inline website/docs/d/lambda_code_signing_config.html.markdown:37]:
```suggestion
* `policies` - List of code signing policies that control the validation failure action for signature mismatch or expiry.
```

**Review by anGie44 — APPROVED** [review]
Niice, let's get this pup in 🚀 🚀

**Review by bflad — APPROVED** [review]
Looks good from me too, minor stuff we can handle later.

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccDataSourceAWSLambdaCodeSigningConfig_PolicyConfigId (12.24s)
--- PASS: TestAccDataSourceAWSLambdaCodeSigningConfig_basic (12.34s)
--- PASS: TestAccDataSourceAWSLambdaCodeSigningConfig_Description (12.50s)

--- PASS: TestAccDataSourceAWSLambdaFunction_environment (36.94s)
--- PASS: TestAccDataSourceAWSLambdaFunction_alias (47.40s)
--- PASS: TestAccDataSourceAWSLambdaFunction_basic (50.94s)
--- PASS: TestAccDataSourceAWSLambdaFunction_version (57.76s)
--- PASS: TestAccDataSourceAWSLambdaFunction_layers (67.52s)

--- PASS: TestAccDataSourceAWSLambdaLayerVersion_version (14.99s)
--- PASS: TestAccDataSourceAWSLambdaLayerVersion_basic (19.91s)
--- PASS: TestAccDataSourceAWSLambdaLayerVersion_runtime (28.28s)

--- PASS: TestAccAWSLambdaCodeSigningConfig_basic (15.36s)
--- PASS: TestAccAWSLambdaCodeSigningConfig_UpdatePolicy (23.63s)
--- PASS: TestAccAWSLambdaCodeSigningConfig_UpdatePublishers (24.60s)

--- PASS: TestAccAWSLambdaFunction_basic (71.96s)
--- PASS: TestAccAWSLambdaFunction_codeSigningConfig (1046.52s)
--- PASS: TestAccAWSLambdaFunction_concurrency (1040.52s)
--- PASS: TestAccAWSLambdaFunction_concurrencyCycle (120.33s)
--- PASS: TestAccAWSLambdaFunction_DeadLetterConfig (1040.64s)
--- PASS: TestAccAWSLambdaFunction_DeadLetterConfigUpdated (122.58s)
--- PASS: TestAccAWSLambdaFunction_disablePublish (1025.85s)
--- PASS: TestAccAWSLambdaFunction_disappears (1003.11s)
--- PASS: TestAccAWSLambdaFunction_EmptyVpcConfig (57.58s)
--- PASS: TestAccAWSLambdaFunction_enablePublish (1217.32s)
--- PASS: TestAccAWSLambdaFunction_encryptedEnvVariables (135.35s)
--- PASS: TestAccAWSLambdaFunction_envVariables (161.05s)
--- PASS: TestAccAWSLambdaFunction_expectFilenameAndS3Attributes (19.21s)
--- PASS: TestAccAWSLambdaFunction_FileSystemConfig (1903.63s)
--- PASS: TestAccAWSLambdaFunction_KmsKeyArn_NoEnvironmentVariables (108.09s)
--- PASS: TestAccAWSLambdaFunction_Layers (924.06s)
--- PASS: TestAccAWSLambdaFunction_LayersUpdate (1050.35s)
--- PASS: TestAccAWSLambdaFunction_localUpdate (1052.39s)
--- PASS: TestAccAWSLambdaFunction_localUpdate_nameOnly (1053.93s)
--- PASS: TestAccAWSLambdaFunction_nilDeadLetterConfig (1150.88s)
--- PASS: TestAccAWSLambdaFunction_runtimes (405.97s)
--- PASS: TestAccAWSLambdaFunction_s3 (39.31s)
--- PASS: TestAccAWSLambdaFunction_s3Update_basic (64.29s)
--- PASS: TestAccAWSLambdaFunction_s3Update_unversioned (64.31s)
--- PASS: TestAccAWSLambdaFunction_tags (84.06s)
--- PASS: TestAccAWSLambdaFunction_tracingConfig (1032.25s)
--- PASS: TestAccAWSLambdaFunction_UnpublishedCodeUpdate (1211.68s)
--- PASS: TestAccAWSLambdaFunction_versioned (65.28s)
--- PASS: TestAccAWSLambdaFunction_versionedUpdate (1234.96s)
--- PASS: TestAccAWSLambdaFunction_VPC (788.22s)
--- PASS: TestAccAWSLambdaFunction_VPC_withInvocation (966.94s)
--- PASS: TestAccAWSLambdaFunction_VpcConfig_ProperIamDependencies (962.36s)
--- PASS: TestAccAWSLambdaFunction_VPCRemoval (1375.17s)
--- PASS: TestAccAWSLambdaFunction_VPCUpdate (2505.36s)

--- PASS: TestAccAWSLambdaLayerVersion_basic (14.36s)
--- PASS: TestAccAWSLambdaLayerVersion_compatibleRuntimes (19.15s)
--- PASS: TestAccAWSLambdaLayerVersion_licenseInfo (25.70s)
--- PASS: TestAccAWSLambdaLayerVersion_description (29.40s)
--- PASS: TestAccAWSLambdaLayerVersion_s3 (29.94s)
--- PASS: TestAccAWSLambdaLayerVersion_update (30.76s)
```

File: aws/data_source_aws_lambda_code_signing_config_test.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccDataSourceAWSLambdaCodeSigningConfig_basic(t *testing.T) {
+	dataSourceName := "data.aws_lambda_code_signing_config.test"
+	resourceName := "aws_lambda_code_signing_config.test"
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
```
[the remark this answers is not in the document]
Comment (pjaudiomv) [reply aws/data_source_aws_lambda_code_signing_config_test.go:13]: I am on gov cloud and am not using this data source however i am getting errors on all my pipelines after release. ref https://github.com/hashicorp/terraform-provider-aws/issues/16398

**Comment by rayjlinden** [comment]
I'm trying to track down what broke our development environment.  I am seeing this errors in the logs:
```
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: ---[ REQUEST POST-SIGN ]-----------------------------
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: GET /2020-06-30/functions/webhook_lambda/code-signing-config HTTP/1.1
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Host: localstack:4566
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: User-Agent: aws-sdk-go/1.35.34 (go1.14.5; linux; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.13.5 (+https://www.terraform.io)
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Authorization: AWS4-HMAC-SHA256 Credential=fake/20201124/us-west-2/lambda/aws4_request, SignedHeaders=content-type;host;x-amz-date, Signature=a971821d4b5c934174283626eaa12a1b49756a05d68e57183af680f7099e424c
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Content-Type: application/json
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: X-Amz-Date: 20201124T054630Z
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Accept-Encoding: gzip
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5:
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5:
2020-11-24T05:46:30.489Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: -----------------------------------------------------
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: 2020/11/24 05:46:30 [DEBUG] [aws-sdk-go] DEBUG: Response lambda/GetFunctionCodeSigningConfig Details:
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: ---[ RESPONSE ]--------------------------------------
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: HTTP/1.1 404
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Connection: close
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Content-Length: 232
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Access-Control-Allow-Headers: authorization,content-type,content-md5,cache-control,x-amz-content-sha256,x-amz-date,x-amz-security-token,x-amz-user-agent,x-amz-target,x-amz-acl,x-amz-version-id,x-localstack-target,x-amz-tagging
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Access-Control-Allow-Methods: HEAD,GET,PUT,POST,DELETE,OPTIONS,PATCH
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Access-Control-Allow-Origin: *
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Access-Control-Expose-Headers: x-amz-version-id
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Content-Type: text/html; charset=utf-8
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Date: Tue, 24 Nov 2020 05:46:30 GMT
2020-11-24T05:46:30.620Z [DEBUG] plugin.terraform-provider-aws_v3.17.0_x5: Server: hypercorn-h11
```

The key thing in our dev environment is that I am using localstack.  Was this tested against localstack?

It appears Localstack does not implement these new endpoints.  Is there a way to disable the checks for code signing configs?

The following already exist:
   skip_credentials_validation = true
  skip_metadata_api_check     = true
  skip_requesting_account_id  = true

Is there a 
  skip_code_signing_checks = true

Or something like it?

**Pull request merged**